src/main/kotlin/com/openlattice/codex/controllers/CodexController.kt

```diff
@@ -38,98 +32,51 @@ import javax.inject.Inject
-
-        textingExecutor.execute {
-            Stream.generate { twilioQueue.take() }.forEach { (organizationId, messageEntitySetId, messageContents, toPhoneNumber) ->
-                //Not very efficient.
-                val phone = organizations.getOrganization(organizationId)!!.smsEntitySetInfo
-                        .flatMap { (phoneNumber, _, entitySetIds, _) -> entitySetIds.map { it to phoneNumber } }
-                        .toMap()
-                        .getValue(messageEntitySetId)
-
-                if (phone == "") {
-                    throw BadRequestException("No source phone number set for organization!")
-                }
-                val message = Message.creator(PhoneNumber(toPhoneNumber), PhoneNumber(phone), messageContents)
-                        .setStatusCallback(URI.create("https://api.openlattice.com/datastore/kodex/status")).create()
-                pendingTexts[message.sid] = message
-                processQueueEntry(message, organizationId)
-            }
-        }
-    }
+    private val twilioQueue = HazelcastQueue.TWILIO.getQueue(hazelcastInstance)
 
     @Timed
     @RequestMapping(path = ["", "/"], method = [RequestMethod.POST])
     override fun sendOutgoingText(@RequestBody contents: MessageRequest) {
+        ensureWriteAccess(AclKey(contents.messageEntitySetId))
+        contents.senderId = Principals.getCurrentUser().id
         twilioQueue.put(contents)
     }
 
-    override fun receiveIncomingText(@RequestBody message: Message) {
-
+    @Timed
+    @RequestMapping(path = [INCOMING + ORG_ID_PATH], method = [RequestMethod.POST], consumes = [MediaType.APPLICATION_FORM_URLENCODED_VALUE])
+    fun receiveIncomingText(@PathVariable(ORG_ID) organizationId: UUID, request: HttpServletRequest) {
+        codexService.processIncomingMessage(organizationId, request)
     }
 
     @Timed
-    @RequestMapping(path = [CodexApi.STATUS], method = [RequestMethod.POST])
-    override fun listenForTextStatus(@RequestBody message: Message) {
-        if (message.status == Message.Status.FAILED || message.status == Message.Status.UNDELIVERED) {
-            println("Message not received or even failed to send!!! ")
-        } else {
-            pendingTexts.remove(message.sid)
-        }
-    }
+    @RequestMapping(path = [INCOMING + ORG_ID_PATH + STATUS], method = [RequestMethod.POST])
+    fun listenForTextStatus(@PathVariable(ORG_ID) organizationId: UUID, request: HttpServletRequest) {
 
-    fun processQueueEntry(message: Message, organizationId: UUID) {
-        createEntitiesFromMessage(message, organizationId, peopleUUID)
-        createEntitiesFromMessage(message, organizationId, contactUUID)
-        createEntitiesFromMessage(message, organizationId, messagesUUID)
-        createEntitiesFromMessage(message, organizationId, toUUID)
-        createEntitiesFromMessage(message, organizationId, fromUUID)
+        val messageId = request.getParameter(CodexConstants.Request.SID.parameter)
```

Comment:
should we do a permission/ownership check here?

Reply:
It's a bit tricky since all the requests will be coming from Twilio -- I can look into whether we can manually set some header for the Twilio requests that we can use for an auth check here or something

Reply:
Actually we can probably do that here using the twilio session, but the other endpoint for receiving incoming messages will be tricky

Reply:
nvm there appears to be a way... :D